skrifa/src/outline/autohint/style.rs

```diff
@@ -51,6 +51,121 @@ impl Default for GlyphStyle {
+///
+/// Also keeps track of the styles that are actually used so we can allocate
+/// an appropriately sized metrics array.
+#[derive(Debug)]
+pub(crate) struct GlyphStyleMap {
+    /// List of styles, indexed by glyph id.
+    styles: Vec<GlyphStyle>,
+    /// Maps an actual style class index into a compacted index for the
+    /// metrics table.
+    ///
+    /// Uses `0xFF` to signify unused styles.
+    metrics_map: [u8; MAX_STYLES],
+    /// Number of metrics styles in use.
+    metrics_count: u8,
+}
+
+impl GlyphStyleMap {
+    /// Computes a new glyph style map for the given glyph count and character
+    /// map.
+    ///
+    /// Roughly based on <https://gitlab.freedesktop.org/freetype/freetype/-/blob/57617782464411201ce7bbc93b086c1b4d7d84a5/src/autofit/afglobal.c#L126>
+    pub fn new(glyph_count: u32, charmap: &Charmap) -> Self {
+        let mut map = Self {
+            styles: vec![GlyphStyle::default(); glyph_count as usize],
+            metrics_map: [0xFF; MAX_STYLES],
+            metrics_count: 0,
+        };
+        // TODO: Step 1: compute styles for glyphs covered by OpenType features
+        // --
+        // Step 2: compute styles for glyphs contained in the cmap
+        // cmap entries are sorted so we keep track of the most recent range to
+        // avoid a binary search per character
+        let mut last_range: Option<(usize, StyleRange)> = None;
+        for (ch, gid) in charmap.mappings() {
+            let Some(style) = map.styles.get_mut(gid.to_u32() as usize) else {
+                continue;
+            };
+            if !style.is_unassigned() {
+                continue;
+            }
+            // Charmaps enumerate in order so we're likely to encounter at least
+            // a few codepoints in the same range.
+            if let Some(last) = last_range {
+                if last.1.contains(ch) {
+                    *style = last.1.style;
+                    continue;
+                }
+            }
+            let ix = match STYLE_RANGES.binary_search_by(|x| x.first.cmp(&ch)) {
+                Ok(i) => i,
+                Err(i) => i.saturating_sub(1),
+            };
+            let Some(range) = STYLE_RANGES.get(ix).copied() else {
+                continue;
+            };
+            if range.contains(ch) {
+                *style = range.style;
+                if let Some(style_ix) = range.style.style_index() {
+                    let style_ix = style_ix as usize;
+                    if map.metrics_map[style_ix] == 0xFF {
+                        // This the first time we've seen this style so record
+                        // it in the metrics map
+                        map.metrics_map[style_ix] = map.metrics_count;
+                        map.metrics_count += 1;
+                    }
+                }
+                last_range = Some((ix, range));
+            }
+        }
+        // TODO: Step 3: compute script based coverage
```

Comment:
are these actual todos? or were they once?

Reply:
These are new and just markers for where I need to add "shaping" support, hopefully next week.